### PR TITLE
add shiny and support for websockets

### PR DESCRIPTION
**Pull request by littledice**

### CODE REVIEW

File: nginx.conf [hunk 1]
```diff
@@ -33,5 +33,10 @@ http {
 
     #gzip  on;
 
+    map $http_upgrade $connection_upgrade {
+        default upgrade;
+        ''      close;
+    }
```
Comment (MartinEden) [inline nginx.conf]: Could you add a comment documenting what this does? It's clear as mud to me!

File: nginx.montagu.conf [hunk 2]
```diff
@@ -49,6 +49,23 @@ server {
...
     }
+
+    location /shiny/ {
+        rewrite ^/shiny/(.*)$ /$1 break;
+        
+        proxy_pass http://shiny_proxy;
+        proxy_redirect http://shiny_proxy/ $scheme://$host/shiny/;
+        proxy_http_version 1.1;
+        proxy_set_header Upgrade $http_upgrade;
+        proxy_set_header Connection $connection_upgrade;
```
Comment (MartinEden) [inline nginx.montagu.conf]: I guess the bit above that I didn't understand relates to this... but I don't know what the Upgrade and Connection headers do.

File: nginx.montagu.conf [hunk 3]
```diff
@@ -49,6 +49,23 @@ server {
...
+        proxy_read_timeout 20d;
+        proxy_buffering off;
+
+        proxy_set_header    Host $host;
+        proxy_set_header    X-Real-IP $remote_addr;
+        proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header    X-Forwarded-Proto $scheme;
```
Comment (MartinEden) [inline nginx.montagu.conf]: A brief comment explaining why we need to forward these three headers would be helpful too :)

**Comment by littledice** [comment]
Very reasonable requests, this stuff is fairly unreadable! I've added comments, maybe @richfitz could confirm that they match his understanding of this, as he is the one who got this working initially.

**Comment by richfitz** [comment]
This link should be added somewhere (it was there in some previous version but looks like I have since deleted the comment when tidying up false starts)

https://support.rstudio.com/hc/en-us/articles/213733868-Running-Shiny-Server-with-a-Proxy

Unfortunately there's very little about the _why_ for the various options.  I think the upgrade one prevents anything upgrading http 1.2 to http 2 because that breaks ... _something_

**Comment by richfitz** [comment]
This looks to be get towards the "why": http://nginx.org/en/docs/http/websocket.html

**Comment by littledice** [comment]
Cool, have added a link to that RStudio doc. I'll link to the nginx one too.

**Comment by littledice** [comment]
Ok, does this seem clearer @MartinEden?

**Pull request merged**